### PR TITLE
first iteration for import of data values

### PR DESCRIPTION
**Pull request by ngermeau**
I know that I should make api call to verify that the data are in dhis2 in the import spec.

### CODE REVIEW

File: lib/dhis2_cli/models/import/data_values_file.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  attr_reader :filepath, :headers, :lines
+
+  def initialize(filepath, headers, lines)
+    @filepath = filepath
+    @headers = headers
+    @lines = lines
+  end
+
+  def data_element_headers
+    @data_element_headers ||= @headers.select { |header| header.start_with?('de') }
```
Comment (borisrorsvort) [inline lib/dhis2_cli/models/import/data_values_file.rb]: The start_with is not auto explaining and need domain knowledge. maybe extract to private method with meaningful name

File: lib/dhis2_cli/models/import/data_values_file.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def data_values_lines
+    @data_values_lines ||= @lines.each_with_index.map do |line, idx|
```
Comment (borisrorsvort) [inline lib/dhis2_cli/models/import/data_values_file.rb]: is idx index ? Then maybe rename to that

File: lib/dhis2_cli/models/import/data_values_file.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      OpenStruct.new Hash[@headers.zip line].merge(line_number: idx + 1)
+    end
+  end
+
+  def data_values
+    data_values_lines.map do |data_value_line|
```
Comment (borisrorsvort) [inline lib/dhis2_cli/models/import/data_values_file.rb]: you can use flap_map I think. Then remove flatten at the end

File: spec/models/data_values_file_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
+require_relative '../../lib/dhis2_cli'
+
+describe DataValuesFile do
+  let(:filepath) { './spec/fixtures/data_values_template.csv' }
+  let(:csv) { CSV.read(filepath, 'r') }
+  let(:data_values_file) { DataValuesFile.new(filepath, csv.first, csv.drop(1)) }
+
+  describe 'data_element_headers' do
+    it 'does return the data elements headers' do
```
Comment (borisrorsvort) [inline spec/models/data_values_file_spec.rb]: «returns» instead of «does return»

File: spec/models/data_values_file_spec.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      expect(data_values_file.data_element_headers).to eq %w(de_2016_abc321 de_2014Q1_xyz321)
+    end
+  end
+
+  describe 'data_values_lines' do
+    it 'does return the list of data values with the values and the org unit id' do
```
Comment (borisrorsvort) [inline spec/models/data_values_file_spec.rb]: Same here

File: spec/models/data_values_file_spec.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      expected_data_values_lines = [OpenStruct.new(external_path: nil,
+                                                   dhis2_path: '/pM/Jk/ls/lvPW1gVZxnz',
```
Comment (borisrorsvort) [inline spec/models/data_values_file_spec.rb]: Extract structs to let vars?

File: spec/helpers/data_values_helper_spec.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+require_relative '../../lib/dhis2_cli'
+
+describe DataValuesHelper do
+  describe 'import' do
+    let(:filepath) { './spec/fixtures/data_values_integration_template.csv' }
+    let(:csv) { CSV.read(filepath, 'r') }
+    let(:data_values_file) { DataValuesFile.new(filepath, csv.first, csv.drop(1)) }
+    let(:uri) { 'https://admin:district@play.dhis2.org/demo' }
+    let(:data_values_helper) { DataValuesHelper.new(uri) }
```
Comment (borisrorsvort) [inline spec/helpers/data_values_helper_spec.rb:9]: You can use `subject {…}` instead

File: lib/dhis2_cli/models/import/data_values_file.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      data_element_headers.map do |data_element|
+        data_value = Hash[[:period, :dataElement].zip data_element.split('_').drop(1)]
+        data_value[:value] = data_value_line[data_element]
+        data_value[:orgUnit] = data_value_line[:dhis2_path].split('/').last
```
Comment (borisrorsvort) [inline lib/dhis2_cli/models/import/data_values_file.rb]: If I'm not mistaken `path` start with `/`. I know you seem to only take the last but shouldn’t we remove the first blank result?

File: spec/models/data_values_file_spec.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                                                   level_1: 'Belgique',
+                                                   level_2: 'Bruxelles',
+                                                   level_3: 'Watermael',
+                                                   facility_name: 'Hive5',
+                                                   de_2016_abc321: '324',
```
Comment (borisrorsvort) [inline spec/models/data_values_file_spec.rb]: Is it normal that numeral values are strings here?

File: lib/dhis2_cli/models/import/data_values_file.rb (hunk 2 above)
Comment (borisrorsvort) [inline lib/dhis2_cli/models/import/data_values_file.rb]: You can also use `map.with_index` at your convenience. Dunno about perf

**Review by borisrorsvort — CHANGES_REQUESTED** [review]
Mostly nitpicking, nice job! Difficult for me now to really test though. Do we have a build status for the pr?

File: lib/dhis2_cli/models/import/data_values_file.rb (hunk 2 above)
Comment (ngermeau) [reply lib/dhis2_cli/models/import/data_values_file.rb]: I think this is rails method no ?

File: lib/dhis2_cli/models/import/data_values_file.rb (hunk 8 above)
Comment (ngermeau) [reply lib/dhis2_cli/models/import/data_values_file.rb]: yep completely right

File: lib/dhis2_cli/models/import/data_values_file.rb (hunk 2 above)
Comment (ngermeau) [reply lib/dhis2_cli/models/import/data_values_file.rb]: it's a rails method, but in ruby you can do arr.map.with_index for simplier syntax

File: spec/models/data_values_file_spec.rb (hunk 6 above)
Comment (ngermeau) [reply spec/models/data_values_file_spec.rb]: only used once so, not really necessary and I don't think it add clarity.

File: spec/models/data_values_file_spec.rb (hunk 9 above)
Comment (ngermeau) [reply spec/models/data_values_file_spec.rb]: yep

File: spec/helpers/data_values_helper_spec.rb (hunk 7 above)
Comment (ngermeau) [reply spec/helpers/data_values_helper_spec.rb:9]: thanks. learning every day.

**Review by borisrorsvort — APPROVED** [review]

**Comment by borisrorsvort** [comment]
@thejax lgtm

**Pull request closed without merging**